### PR TITLE
[CM-1633] Implement Artifact Update part

### PR DESCRIPTION
**Pull request by yaricom**
### comet-java-client
- Exposed variety of `Asset` flavors to the public API 
- Implemented mechanism to update existing artifact using `DownloadedArtifact` returned by `LoggedArtifact#download()`

### comet-examples
- Implemented example of the artifact update in the `ArtifactExample`

### CODE REVIEW

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java [hunk 1]
```diff
@@ -22,15 +23,14 @@
     private String fileExtension;
 
     @ToString.Include
-    String fileName;
+    String logicalPath;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java:26]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 2]
```diff
@@ -41,7 +42,7 @@
     @Getter
     @Setter
     @ToString.Include
-    private String assetType;
+    private AssetType assetType;
     @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:46]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java [hunk 3]
```diff
@@ -14,19 +21,67 @@
 @ToString(callSuper = true)
 public class ArtifactAssetImpl extends AssetImpl implements ArtifactAsset {
     String artifactVersionId;
+    Long fileSize;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java:24]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java [hunk 4]
```diff
@@ -22,15 +23,14 @@
...
     @ToString.Include
     AssetType type;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java:28]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java:26]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java [hunk 5]
```diff
@@ -1,32 +1,100 @@
...
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Optional;
 
 /**
  * Describes artifact's asset data.
  */
-@Data
-@NoArgsConstructor
 @EqualsAndHashCode(callSuper = true)
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java:19]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java [hunk 6]
```diff
@@ -1,32 +1,100 @@
...
 @ToString(callSuper = true)
-public class ArtifactAssetImpl extends AssetImpl implements ArtifactAsset {
+public final class ArtifactAssetImpl extends RemoteAssetImpl implements ArtifactAsset {
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java:22]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java [hunk 7]
```diff
@@ -1,32 +1,100 @@
...
+    @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/ArtifactAssetImpl.java:23]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java [hunk 8]
```diff
@@ -1,37 +1,75 @@
...
     @ToString.Include
-    private File file;
-    private byte[] fileLikeData;
+    @Getter
+    @Setter
+    private File rawFile;
+    @Getter
+    @Setter
+    private byte[] rawFileLikeData;
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java:27]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java [hunk 9]
```diff
@@ -1,37 +1,75 @@
...
     private String fileExtension;
 
     @ToString.Include
-    String fileName;
+    @Getter
+    @Setter
+    String logicalPath;
     @ToString.Include
+    @Getter
+    @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java:37]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import java.time.Instant;
+import java.util.Optional;
+
+/**
+ * Implementation of the {@link LoggedExperimentAsset}.
+ */
+@EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true)
+public final class LoggedExperimentAssetImpl extends RemoteAssetImpl implements LoggedExperimentAsset {
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java:18]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    @Setter
+    private String assetId;
+    @Setter
+    private boolean remote;
+    @Setter
+    private Instant createdAt;
+    @Getter
+    @Setter
+    private Long fileSize;
+    @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java:28]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import lombok.ToString;
+import ml.comet.experiment.asset.LoggedExperimentAsset;
+
+import java.time.Instant;
+import java.util.Optional;
+
+/**
+ * Implementation of the {@link LoggedExperimentAsset}.
+ */
+@EqualsAndHashCode(callSuper = true)
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java:15]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/RemoteAssetImpl.java [hunk 13]
```diff
@@ -1,17 +1,25 @@
...
 
 /**
  * Describes remote asset data.
  */
-@Data
-@NoArgsConstructor
 @EqualsAndHashCode(callSuper = true)
 public class RemoteAssetImpl extends AssetImpl implements RemoteAsset {
-    private URI link;
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/RemoteAssetImpl.java:17]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/RemoteAssetImpl.java [hunk 14]
```diff
@@ -1,17 +1,25 @@
...
+    @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/RemoteAssetImpl.java:18]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+@ToString(callSuper = true)
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java:16]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java:15]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/DownloadedArtifactImpl.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+import static ml.comet.experiment.impl.resources.LogMessages.CONFLICTING_ARTIFACT_ASSET_NAME;
+import static ml.comet.experiment.impl.resources.LogMessages.FAILED_TO_SET_ARTIFACT_VERSION_LEQ_THAN_CURRENT;
+import static ml.comet.experiment.impl.resources.LogMessages.getString;
+
+/**
+ * The implementation of the {@link ml.comet.experiment.artifact.Artifact} to represent artifact downloaded from the
+ * Comet by {@link ml.comet.experiment.artifact.LoggedArtifact#download(Path, AssetOverwriteStrategy)} method.
+ */
+public final class DownloadedArtifactImpl extends ArtifactImpl implements DownloadedArtifact {
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/DownloadedArtifactImpl.java]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 17]
```diff
@@ -41,7 +42,7 @@
...
     @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:47]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java [hunk 18]
```diff
@@ -1,37 +1,83 @@
...
-@NoArgsConstructor
 @ToString(onlyExplicitlyIncluded = true)
 public class AssetImpl implements Asset {
     @ToString.Include
-    private File file;
-    private byte[] fileLikeData;
+    @Getter
+    @Setter
+    private File rawFile;
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/AssetImpl.java:24]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+public final class LoggedExperimentAssetImpl extends RemoteAssetImpl implements LoggedExperimentAsset {
+    @Setter
+    private String assetId;
+    @Setter
+    private boolean remote;
+    @Setter
+    private Instant createdAt;
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/LoggedExperimentAssetImpl.java:24]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/DownloadedArtifactImpl.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+    private final Logger logger = LoggerFactory.getLogger(DownloadedArtifact.class);
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/DownloadedArtifactImpl.java:29]: not relevant

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 21]
```diff
@@ -61,8 +61,9 @@
     private String artifactId;
     @Setter
     private String artifactVersionId;
-
+    @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:64]: @sonatype-lift ignore

**Review by Lothiraldan — APPROVED** [review]
Looks good to me!

**Pull request merged**